lucene/core/src/java/org/apache/lucene/search/PhraseQuery.java

```diff
@@ -284,6 +284,16 @@ public Query rewrite(IndexReader reader) throws IOException {
     }
   }
 
+  @Override
+  public void visit(QueryVisitor visitor) {
+    for (Term term : terms) {
+      QueryVisitor v = visitor.getSubVisitor(BooleanClause.Occur.MUST, this);
```

Comment:
don't pull this more than once?

Reply:
The advantage of pulling multiple times is the combination with the null check, meaning we can short-circuit; it's particularly useful on something like BooleanQuery or DisjunctionQuery, where if you're say looking to see if there's a single instance of a particular term or query type, you can stop looking as soon as you find it rather than having to continue traversing through all the possibilities

Reply:
[I think you need to read this. Maybe twice.](https://en.wikipedia.org/wiki/Program_optimization)

Reply:
I took a shortcut :p

More seriously, I came across issues like this in Luwak when dealing with super-massive queries, but I suppose there are other ways around it, like throwing an exception or whatever.  I can change it back to returning an empty visitor if you really don't like it.

Reply:
you can enforce this by adding a boolean return to `consumeTerm` that we can respect. that would be a clear API to short-circuit.

Reply:
I agree with Simon. I find it counter intuitive to get the sub-visitor for the same query more than once. if we do need that short-circuiting maybe it can be made more explicit in the API.

Probably because I am not yet very familiar with the change here, but I have a hard time following where things happen here: we pull the sub-visitor from the same parent visitor multiple times, yet what gets returned may be null at some point. Does that mean that the parent visitor holds state that changes as we consume terms through its sub-visitors?